dashboard/app/controllers/api/v1/assessments_controller.rb

```diff
@@ -0,0 +1,199 @@
+
+        # Get the response which will be stringified JSON.
+        response = last_attempt.try(:level_source).try(:data)
+
+        next unless response
+
+        # Parse the response string into an object.
+        response_parsed = JSON.parse(response)
+
+        # Summarize some key data.
+        multi_count = 0
+        multi_count_correct = 0
+
+        # And construct a listing of all the individual levels and their results.
+        level_results = []
+
+        level_group.levels.each do |level|
+          if level.is_a? Multi
+            multi_count += 1
+          end
+
+          level_response = response_parsed[level.id.to_s]
+
+          level_result = {}
+
+          if level_response
+            case level
+            when TextMatch, FreeResponse
+              student_result = level_response["result"]
+              level_result[:student_result] = student_result
+              level_result[:status] = "free_response"
+            when Multi
+              answer_indexes = Multi.find_by_id(level.id).correct_answer_indexes
+              student_result = level_response["result"].split(",").sort.join(",")
+
+              # Convert "0,1,3" to "A, B, D" for teacher-friendly viewing
+              level_result[:student_result] = student_result.split(',').map {|k| Multi.value_to_letter(k.to_i)}.join(', ')
+
+              if student_result == "-1"
+                level_result[:student_result] = ""
+                level_result[:status] = "unsubmitted"
+              elsif student_result == answer_indexes
+                multi_count_correct += 1
+                level_result[:status] = "correct"
+              else
+                level_result[:status] = "incorrect"
+              end
+            end
+          else
+            level_result[:status] = "unsubmitted"
+          end
+
+          level_results << level_result
+        end
+
+        submitted = last_attempt[:submitted]
+        timestamp = last_attempt[:updated_at].to_formatted_s
+
+        responses_by_level_group[script_level.id] = {
+          stage: script_level.stage.localized_title,
+          puzzle: script_level.position,
+          question: level_group.properties["title"],
+          url: build_script_level_url(script_level, section_id: section.id, user_id: student.id),
+          multi_correct: multi_count_correct,
+          multi_count: multi_count,
+          submitted: submitted,
+          timestamp: timestamp,
+          level_results: level_results
+        }
+      end
+
+      # If a student has made responses, add them to the hash
+      if responses_by_level_group.keys.count > 0
+        student_hash[:responses_by_assessment] = responses_by_level_group
+        responses_by_student[student.id] = student_hash
+      end
+    end
+
+    render json: responses_by_student
+  end
+
+  private
+
+  def load_section
+    section = Section.find(params[:section_id])
+    authorize! :read, section
+    section
+  end
```

Comment:
When it comes time to refactor, this looks like a perfect candidate for CanCan's `load_and_authorize_resource`